### ISSUE TITLE
XmlException: An error was encountered when parsing a DateTime value in the XML

### ISSUE THREAD

**Issue by dougwaldron** [dougwaldron/podcast-rewind#75]
Reported via [Sentry](https://d-w.sentry.io/share/issue/ba2c9c9bf06e48df9f35d5a6591530d6/):

Affected page: https://podcast-rewind.azurewebsites.net/Setup?feedUrl=https%3A%2F%2Fdaringfireball.net%2Fthetalkshow%2Frss

Feed URL: `https://daringfireball.net/thetalkshow/rss`

</body>
</html>

Stack trace: 

```
System.Xml.XmlException: Error in line 754 position 43. An error was encountered when parsing a DateTime value in the XML.
  File "SyndicationItem.cs", line 138, in DateTimeOffset SyndicationItem.get_PublishDate()
    throw PublishDateException;
  File "Models/Dto/ViewPodcastEpisodeDto.cs", line 13, col 56, in new ViewPodcastEpisodeDto(SyndicationItem item)
  File "Pages/Setup.cshtml.cs", line 60, col 53, in void SetupModel.LoadData(SyndicationFeed feed)+(SyndicationItem item) => { } [0]
  File "Select.SpeedOpt.cs", line 428, in void SelectIListIterator<TSource, TResult>.Fill(IList<TSource> source, Span<TResult> results, Func<TSource, TResult> func)
    results[i] = func(source[i]);
  File "Select.SpeedOpt.cs", line 408, in TResult[] SelectIListIterator<TSource, TResult>.ToArray()
    Fill(_source, results, _selector);
  File "Buffer.cs", line 32, in new Buffer<TElement>(IEnumerable<TElement> source)
    TElement[] array = iterator.ToArray();
  File "OrderedEnumerable.SpeedOpt.cs", line 29, in List<TElement> OrderedEnumerable<TElement>.ToList()
    Buffer<TElement> buffer = new Buffer<TElement>(_source);
  File "ToCollection.cs", line 24, in List<TSource> Enumerable.ToList<TSource>(IEnumerable<TSource> source)
    if (source == null)
  File "Pages/Setup.cshtml.cs", line 60, col 9, in void SetupModel.LoadData(SyndicationFeed feed)
  File "Pages/Setup.cshtml.cs", line 27, col 9, in async Task<IActionResult> SetupModel.OnGetAsync(string feedUrl, int interval)
  File "ExecutorFactory.cs", line 139, in async Task<object> GenericTaskHandlerMethod.Convert<T>(object taskAsObject)
    return await task;
  File "ExecutorFactory.cs", line 132, in async Task<IActionResult> GenericTaskHandlerMethod.Execute(object receiver, object[] arguments)
    var result = await _thunk(receiver, arguments);
  File "PageActionInvoker.cs", line 274, in async Task PageActionInvoker.InvokeHandlerMethodAsync()
    _result = await executor(_instance, arguments);
  File "PageActionInvoker.cs", line 659, in async Task PageActionInvoker.InvokeNextPageFilterAsync()
    await Next(ref next, ref scope, ref state, ref isCompleted);
  File "PageActionInvoker.cs", line 706, in void PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
    context.ExceptionDispatchInfo?.Throw();
  File "PageActionInvoker.cs", line 633, in Task PageActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
    Rethrow(handlerExecutedContext);
  File "PageActionInvoker.cs", line 83, in async Task PageActionInvoker.InvokeInnerFilterAsync()
    await Next(ref next, ref scope, ref state, ref isCompleted);
  File "ResourceInvoker.cs", line 978, in async Task ResourceInvoker.InvokeNextResourceFilter()+Awaited(?)
    await lastTask;
  File "ResourceInvoker.cs", line 1460, in void ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
    }
  File "ResourceInvoker.cs", line 890, in Task ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
    Rethrow(_resourceExecutedContext!);
  File "ResourceInvoker.cs", line 254, in async Task ResourceInvoker.InvokeFilterPipelineAsync()+Awaited(?)
    await invoker.Next(ref next, ref scope, ref state, ref isCompleted);
  File "ResourceInvoker.cs", line 124, in async Task ResourceInvoker.InvokeAsync()+Logged(?) x 2
    await invoker.InvokeFilterPipelineAsync();
  ?, in async Task SentryTracingMiddleware.InvokeAsync(HttpContext context) x 2
  File "ExceptionHandlerMiddlewareImpl.cs", line 98, in async Task ExceptionHandlerMiddlewareImpl.Invoke(HttpContext context)+Awaited(?)
    await task;
```

**Comment by dougwaldron** [comment]
Well that was fun evening diving into the code and specs!

*TLDR; It looks like the .NET `SyndicationFeed` parser incorrectly throws an exception when the date uses a single digit day.*

---

The RSS feed for [The Talk Show](https://daringfireball.net/thetalkshow/) uses single-digit days (when less than 10). The `SyndicationFeed` parser has no problem with this episode:

```xml
<guid>https://daringfireball.net/thetalkshow/2023/06/17/ep-379</guid>
<pubDate>Sat, 17 Jun 2023 19:22:55 EDT</pubDate>
```

But fails for the next episode:

```xml

<guid>https://daringfireball.net/thetalkshow/2023/06/07/ep-378</guid>
<pubDate>Wed, 7 Jun 2023 20:13:43 EDT</pubDate>
```

with the exception message: 

> An error was encountered when parsing a DateTime value in the XML.

---

Diving into the code... 

When `SyndicationFeed` tries load an RSS feed, it first attempts to [parse publication dates](https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/DateTimeHelper.cs#L17-L21) using the default `DateTimeOffset` parser. If that doesn't work, it uses `DateTimeOffset.TryParseExact` with an [array of possible date formats](https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/DateTimeHelper.cs#L93-L117) based on RFC 822.

The [RSS 2.0 spec](https://www.rssboard.org/rss-specification#optionalChannelElements) does specify that `pubDate` should conform to RFC 822 (except that years should be 4 digits).

[RFC 822](https://www.w3.org/Protocols/rfc822/#z28) in turn specifies that the date portion of the date-time value be formatted like so:

```
date        =  1*2DIGIT month 2DIGIT        ; day month year
                                            ;  e.g. 20 Jun 82
```

where that `1*2DIGIT` [notation indicates](https://www.w3.org/Protocols/rfc822/#z25) at least 1 and at most 2 digits for the day value. So single-digit dates are valid.

But the `parseFormat` array includes no formats with single-digit days. This is a bug.

**Comment by dougwaldron** [comment]
Bug report submitted: https://github.com/dotnet/runtime/issues/99193

**Comment by dougwaldron** [comment]
Pull request submitted: https://github.com/dotnet/runtime/pull/99194

**Comment by dougwaldron** [comment]
I updated the Sentry.io link in the original description since the old one had expired.
